// iron_worker/lib/api_client.js
// Generated by CoffeeScript 1.3.3
(function() {
  var APIClient, iron_core, version, _,
    __hasProp = {}.hasOwnProperty,
    __extends = function(child, parent) { for (var key in parent) { if (__hasProp.call(parent, key)) child[key] = parent[key]; } function ctor() { this.constructor = child; } ctor.prototype = parent.prototype; child.prototype = new ctor(); child.__super__ = parent.prototype; return child; };

  require('pkginfo')(module);

  version = this.version;

  _ = require('underscore');

  iron_core = require('iron_core');

  APIClient = (function(_super) {

    __extends(APIClient, _super);

    APIClient.prototype.AWS_US_EAST_HOST = 'worker-aws-us-east-1.iron.io';

    function APIClient(options) {
      var default_options;
      default_options = {
        scheme: 'https',
        host: this.AWS_US_EAST_HOST,
        port: 443,
        api_version: 2
      };
      APIClient.__super__.constructor.call(this, 'iron', 'worker', options, default_options, ['project_id', 'token', 'api_version']);
    }

    APIClient.prototype.version = function() {
      return "iron_worker_node-" + version + " (" + (APIClient.__super__.version.call(this)) + ")";
    };

    APIClient.prototype.url = function() {
      return APIClient.__super__.url.call(this) + this.options.api_version.toString() + '/';
    };

    APIClient.prototype.headers = function() {
      return _.extend({}, APIClient.__super__.headers.call(this), {
        'Authorization': "OAuth " + this.options.token
      });
    };

    APIClient.prototype.tasks_create = function(code_name, payload, options, cb) {
      var parse_response_bind;
      parse_response_bind = _.bind(this.parse_response, this);
      return this.post("projects/" + this.options.project_id + "/tasks", {
        'tasks': [
          _.extend({
            'code_name': code_name,
            'payload': payload
          }, options)
        ]
      }, function(error, response, body) {
        return parse_response_bind(error, response, body, cb);
      });
    };

    return APIClient;

  })(iron_core.Client);

  module.exports.APIClient = APIClient;

}).call(this);
